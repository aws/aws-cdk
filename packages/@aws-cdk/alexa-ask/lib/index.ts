// Alexa::ASK CloudFormation Resources:
export * from './ask.generated';
export * from './pipeline-actions';
