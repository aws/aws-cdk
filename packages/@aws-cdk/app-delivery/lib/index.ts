export * from './pipeline-deploy-stack-action';
