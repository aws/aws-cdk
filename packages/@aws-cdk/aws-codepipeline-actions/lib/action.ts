import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as events from '@aws-cdk/aws-events';
import { Construct, Lazy } from '@aws-cdk/core';

/**
 * Low-level class for generic CodePipeline Actions.
 *
 * WARNING: this class should not be externally exposed, but is currently visible
 * because of a limitation of jsii (https://github.com/aws/jsii/issues/524).
 *
 * This class will disappear in a future release and should not be used.
 *
 * @experimental
 */
export abstract class Action implements codepipeline.IAction {
  public readonly actionProperties: codepipeline.ActionProperties;
  private _pipeline?: codepipeline.IPipeline;
  private _stage?: codepipeline.IStage;
  private _scope?: Construct;
  private readonly customerProvidedNamespace?: string;
  private readonly namespaceOrToken: string;
  private actualNamespace?: string;
  private variableReferenced = false;

  protected constructor(actionProperties: codepipeline.ActionProperties) {
    this.customerProvidedNamespace = actionProperties.variablesNamespace;
    this.namespaceOrToken = Lazy.stringValue({ produce: () => {
      // make sure the action was bound (= added to a pipeline)
      if (this.actualNamespace !== undefined) {
        return this.customerProvidedNamespace !== undefined
          // if a customer passed a namespace explicitly, always use that
          ? this.customerProvidedNamespace
          // otherwise, only return a namespace if any variable was referenced
          : (this.variableReferenced ? this.actualNamespace : undefined);
      } else {
        throw new Error(`Cannot reference variables of action '${this.actionProperties.actionName}', ` +
          'as that action was never added to a pipeline');
      }
    }});
    this.actionProperties = {
      ...actionProperties,
      variablesNamespace: this.namespaceOrToken,
    };
  }

  public bind(scope: Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig {
    this._pipeline = stage.pipeline;
    this._stage = stage;
    this._scope = scope;

    this.actualNamespace = this.customerProvidedNamespace === undefined
      // default a namespace name, based on the stage and action names
      ? `${stage.stageName}_${this.actionProperties.actionName}_NS`
      : this.customerProvidedNamespace;

    return this.bound(scope, stage, options);
  }

  public onStateChange(name: string, target?: events.IRuleTarget, options?: events.RuleProps) {
    const rule = new events.Rule(this.scope, name, options);
    rule.addTarget(target);
    rule.addEventPattern({
      detailType: [ 'CodePipeline Action Execution State Change' ],
      source: [ 'aws.codepipeline' ],
      resources: [ this.pipeline.pipelineArn ],
      detail: {
        stage: [ this.stage.stageName ],
        action: [ this.actionProperties.actionName ],
      },
    });
    return rule;
  }

  protected variableExpression(variableName: string): string {
    this.variableReferenced = true;
    return `#{${this.namespaceOrToken}.${variableName}}`;
  }

  /**
   * The method called when an Action is attached to a Pipeline.
   * This method is guaranteed to be called only once for each Action instance.
   *
   * @param options an instance of the {@link ActionBindOptions} class,
   *   that contains the necessary information for the Action
   *   to configure itself, like a reference to the Role, etc.
   */
  protected abstract bound(scope: Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig;

  private get pipeline(): codepipeline.IPipeline {
    if (this._pipeline) {
      return this._pipeline;
    } else {
      throw new Error('Action must be added to a stage that is part of a pipeline before using onStateChange');
    }
  }

  private get stage(): codepipeline.IStage {
    if (this._stage) {
      return this._stage;
    } else {
      throw new Error('Action must be added to a stage that is part of a pipeline before using onStateChange');
    }
  }

  /**
   * Retrieves the Construct scope of this Action.
   * Only available after the Action has been added to a Stage,
   * and that Stage to a Pipeline.
   */
  private get scope(): Construct {
    if (this._scope) {
      return this._scope;
    } else {
      throw new Error('Action must be added to a stage that is part of a pipeline first');
    }
  }
}
