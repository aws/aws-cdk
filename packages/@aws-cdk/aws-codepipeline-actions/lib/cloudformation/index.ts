export * from './pipeline-actions';
export * from './stackset-action';
export * from './stackinstances-action';
export * from './stackset-types';
