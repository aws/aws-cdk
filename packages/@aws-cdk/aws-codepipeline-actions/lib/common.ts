import * as codepipeline from '@aws-cdk/aws-codepipeline';

/**
 * The ArtifactBounds that make sense for source Actions -
 * they don't have any inputs, and have exactly one output.
 */
export function sourceArtifactBounds(): codepipeline.ActionArtifactBounds {
  return {
    minInputs: 0,
    maxInputs: 0,
    minOutputs: 1,
    maxOutputs: 1,
  };
}

/**
 * The ArtifactBounds that make sense for deploy Actions -
 * they have exactly one input, and don't produce any outputs.
 */
export function deployArtifactBounds(): codepipeline.ActionArtifactBounds {
  return {
    minInputs: 1,
    maxInputs: 1,
    minOutputs: 0,
    maxOutputs: 0,
  };
}
