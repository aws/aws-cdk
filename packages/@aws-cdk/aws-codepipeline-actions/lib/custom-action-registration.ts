import * as codepipeline from '@aws-cdk/aws-codepipeline';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * The creation attributes used for defining a configuration property
 * of a custom Action.
 */
export interface CustomActionProperty {
  /**
   * The name of the property.
   * You use this name in the `configuration` attribute when defining your custom Action class.
   */
  name: string;

  /**
   * The description of the property.
   *
   * @default the description will be empty
   */
  description?: string;

  /**
   * Whether this property is a key.
   *
   * @default false
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codepipeline-customactiontype-configurationproperties.html#cfn-codepipeline-customactiontype-configurationproperties-key
   */
  key?: boolean;

  /**
   * Whether this property is queryable.
   * Note that only a single property of a custom Action can be queryable.
   *
   * @default false
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codepipeline-customactiontype-configurationproperties.html#cfn-codepipeline-customactiontype-configurationproperties-queryable
   */
  queryable?: boolean;

  /**
   * Whether this property is required.
   */
  required: boolean;

  /**
   * Whether this property is secret,
   * like a password, or access key.
   *
   * @default false
   */
  secret?: boolean;

  /**
   * The type of the property,
   * like 'String', 'Number', or 'Boolean'.
   *
   * @default 'String'
   */
  type?: string;
}

/**
 * Properties of registering a custom Action.
 */
export interface CustomActionRegistrationProps {
  /**
   * The category of the Action.
   */
  category: codepipeline.ActionCategory;

  /**
   * The artifact bounds of the Action.
   */
  artifactBounds: codepipeline.ActionArtifactBounds;

  /**
   * The provider of the Action.
   */
  provider: string;

  /**
   * The version of your Action.
   *
   * @default '1'
   */
  version?: string;

  /**
   * The URL shown for the entire Action in the Pipeline UI.
   */
  entityUrl?: string;

  /**
   * The URL shown for a particular execution of an Action in the Pipeline UI.
   */
  executionUrl?: string;

  /**
   * The properties used for customizing the instance of your Action.
   *
   * @default []
   */
  actionProperties?: CustomActionProperty[];
}

/**
 * The resource representing registering a custom Action with CodePipeline.
 * For the Action to be usable, it has to be registered for every region and every account it's used in.
 * In addition to this class, you should most likely also provide your clients a class
 * representing your custom Action, extending the Action class,
 * and taking the `actionProperties` as properly typed, construction properties.
 */
export class CustomActionRegistration extends Construct {
  constructor(parent: Construct, id: string, props: CustomActionRegistrationProps) {
    super(parent, id);

    new codepipeline.CfnCustomActionType(this, 'Resource', {
      category: props.category,
      inputArtifactDetails: {
        minimumCount: props.artifactBounds.minInputs,
        maximumCount: props.artifactBounds.maxInputs,
      },
      outputArtifactDetails: {
        minimumCount: props.artifactBounds.minOutputs,
        maximumCount: props.artifactBounds.maxOutputs,
      },
      provider: props.provider,
      version: props.version || '1',
      settings: {
        entityUrlTemplate: props.entityUrl,
        executionUrlTemplate: props.executionUrl,
      },
      configurationProperties: props.actionProperties === undefined ? undefined : props.actionProperties.map((ap) => {
        return {
          key: ap.key || false,
          secret: ap.secret || false,
          ...ap,
        };
      }),
    });
  }
}
