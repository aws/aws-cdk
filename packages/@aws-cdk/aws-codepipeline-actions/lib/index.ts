export * from './alexa-ask/deploy-action';
export * from './bitbucket/source-action';
export * from './cloudformation/pipeline-actions';
export * from './codebuild/build-action';
export * from './codecommit/source-action';
export * from './codedeploy/ecs-deploy-action';
export * from './codedeploy/server-deploy-action';
export * from './ecr/source-action';
export * from './ecs/deploy-action';
export * from './github/source-action';
export * from './jenkins/jenkins-action';
export * from './jenkins/jenkins-provider';
export * from './lambda/invoke-action';
export * from './manual-approval-action';
export * from './s3/deploy-action';
export * from './s3/source-action';
export * from './action'; // for some reason, JSII fails building the module without exporting this class
