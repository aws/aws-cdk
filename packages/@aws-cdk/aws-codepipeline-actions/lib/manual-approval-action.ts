import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as sns from '@aws-cdk/aws-sns';
import * as subs from '@aws-cdk/aws-sns-subscriptions';
import * as cdk from '@aws-cdk/core';
import { Action } from './action';

/**
 * Construction properties of the {@link ManualApprovalAction}.
 */
export interface ManualApprovalActionProps extends codepipeline.CommonAwsActionProps {
  /**
   * Optional SNS topic to send notifications to when an approval is pending.
   */
  readonly notificationTopic?: sns.ITopic;

  /**
   * A list of email addresses to subscribe to notifications when this Action is pending approval.
   * If this has been provided, but not `notificationTopic`,
   * a new Topic will be created.
   */
  readonly notifyEmails?: string[];

  /**
   * Any additional information that you want to include in the notification email message.
   */
  readonly additionalInformation?: string;

  /**
   * URL you want to provide to the reviewer as part of the approval request.
   *
   * @default - the approval request will not have an external link
   */
  readonly externalEntityLink?: string;
}

/**
 * Manual approval action.
 */
export class ManualApprovalAction extends Action {
  /**
   * The SNS Topic passed when constructing the Action.
   * If no Topic was passed, but `notifyEmails` were provided,
   * a new Topic will be created.
   */
  private _notificationTopic?: sns.ITopic;
  private readonly props: ManualApprovalActionProps;

  constructor(props: ManualApprovalActionProps) {
    super({
      ...props,
      category: codepipeline.ActionCategory.APPROVAL,
      provider: 'Manual',
      artifactBounds: { minInputs: 0, maxInputs: 0, minOutputs: 0, maxOutputs: 0 },
    });

    this.props = props;
  }

  public get notificationTopic(): sns.ITopic | undefined {
    return this._notificationTopic;
  }

  protected bound(scope: cdk.Construct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig {
    if (this.props.notificationTopic) {
      this._notificationTopic = this.props.notificationTopic;
    } else if ((this.props.notifyEmails || []).length > 0) {
      this._notificationTopic = new sns.Topic(scope, 'TopicResource');
    }

    if (this._notificationTopic) {
      this._notificationTopic.grantPublish(options.role);
      for (const notifyEmail of this.props.notifyEmails || []) {
        this._notificationTopic.addSubscription(new subs.EmailSubscription(notifyEmail));
      }
    }

    return {
      configuration: undefinedIfAllValuesAreEmpty({
        NotificationArn: this._notificationTopic?.topicArn,
        CustomData: this.props.additionalInformation,
        ExternalEntityLink: this.props.externalEntityLink,
      }),
    };
  }
}

function undefinedIfAllValuesAreEmpty(object: object): object | undefined {
  return Object.values(object).some(v => v !== undefined) ? object : undefined;
}
