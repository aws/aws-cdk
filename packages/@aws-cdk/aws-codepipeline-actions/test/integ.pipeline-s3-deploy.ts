import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import * as cpactions from '../lib';

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-s3-deploy');

const bucket = new s3.Bucket(stack, 'PipelineBucket', {
  versioned: true,
  removalPolicy: cdk.RemovalPolicy.DESTROY,
});
const sourceOutput = new codepipeline.Artifact('SourceArtifact');
const sourceAction = new cpactions.S3SourceAction({
  actionName: 'Source',
  output: sourceOutput,
  bucket,
  bucketKey: 'key',
});

const deployBucket = new s3.Bucket(stack, 'DeployBucket', {});

new codepipeline.Pipeline(stack, 'Pipeline', {
  stages: [
    {
      stageName: 'Source',
      actions: [sourceAction],
    },
    {
      stageName: 'Deploy',
      actions: [
        new cpactions.S3DeployAction({
          actionName: 'DeployAction',
          input: sourceOutput,
          bucket: deployBucket,
          accessControl: s3.BucketAccessControl.PUBLIC_READ,
          cacheControl: [
            cpactions.CacheControl.setPublic(),
            cpactions.CacheControl.maxAge(cdk.Duration.hours(12)),
          ],
        }),
      ],
    },
  ],
});

app.synth();
