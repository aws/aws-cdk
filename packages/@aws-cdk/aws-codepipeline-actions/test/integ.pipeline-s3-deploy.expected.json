{
  "Resources": {
    "PipelineBucketB967BD35": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "VersioningConfiguration": {
          "Status": "Enabled"
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "DeployBucket67E2C076": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "PipelineRoleD68726F7": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineRoleDefaultPolicyC7A05455": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineBucketB967BD35",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineBucketB967BD35",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineSourceCodePipelineActionRoleC6F9E7F5",
                  "Arn"
                ]
              }
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "PipelineDeployDeployActionCodePipelineActionRole1C288A60",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": [
          {
            "Ref": "PipelineRoleD68726F7"
          }
        ]
      }
    },
    "PipelineC660917D": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineRoleD68726F7",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1"
                },
                "Configuration": {
                  "S3Bucket": {
                    "Ref": "PipelineBucketB967BD35"
                  },
                  "S3ObjectKey": "key"
                },
                "Name": "Source",
                "OutputArtifacts": [
                  {
                    "Name": "SourceArtifact"
                  }
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "PipelineSourceCodePipelineActionRoleC6F9E7F5",
                    "Arn"
                  ]
                },
                "RunOrder": 1
              }
            ],
            "Name": "Source"
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1"
                },
                "Configuration": {
                  "BucketName": {
                    "Ref": "DeployBucket67E2C076"
                  },
                  "Extract": "true",
                  "CannedACL": "public-read",
                  "CacheControl": "public, max-age=43200"
                },
                "InputArtifacts": [
                  {
                    "Name": "SourceArtifact"
                  }
                ],
                "Name": "DeployAction",
                "RoleArn": {
                  "Fn::GetAtt": [
                    "PipelineDeployDeployActionCodePipelineActionRole1C288A60",
                    "Arn"
                  ]
                },
                "RunOrder": 1
              }
            ],
            "Name": "Deploy"
          }
        ],
        "ArtifactStore": {
          "Location": {
            "Ref": "PipelineBucketB967BD35"
          },
          "Type": "S3"
        }
      },
      "DependsOn": [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7"
      ]
    },
    "PipelineSourceCodePipelineActionRoleC6F9E7F5": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineSourceCodePipelineActionRoleDefaultPolicy2D565925": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineBucketB967BD35",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineBucketB967BD35",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineBucketB967BD35",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineBucketB967BD35",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineSourceCodePipelineActionRoleDefaultPolicy2D565925",
        "Roles": [
          {
            "Ref": "PipelineSourceCodePipelineActionRoleC6F9E7F5"
          }
        ]
      }
    },
    "PipelineDeployDeployActionCodePipelineActionRole1C288A60": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineDeployDeployActionCodePipelineActionRoleDefaultPolicyE194961B": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "DeployBucket67E2C076",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "DeployBucket67E2C076",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": "s3:PutObjectAcl",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "DeployBucket67E2C076",
                        "Arn"
                      ]
                    },
                    "/*"
                  ]
                ]
              }
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineBucketB967BD35",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineBucketB967BD35",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineDeployDeployActionCodePipelineActionRoleDefaultPolicyE194961B",
        "Roles": [
          {
            "Ref": "PipelineDeployDeployActionCodePipelineActionRole1C288A60"
          }
        ]
      }
    }
  }
}